File: app/src/main/java/com/ladsoft/bakingapp/di/module/ModelModule.java
package com.ladsoft.bakingapp.di.module;

import com.ladsoft.bakingapp.data.database.repository.DatabaseIngredientRepository;
import com.ladsoft.bakingapp.data.database.repository.DatabaseRecipeRepository;
import com.ladsoft.bakingapp.data.database.repository.DatabaseStepRepository;
import com.ladsoft.bakingapp.data.repository.RecipeRepository;
import com.ladsoft.bakingapp.manager.SessionManager;
import com.ladsoft.bakingapp.mvp.RecipeMvp;
import com.ladsoft.bakingapp.mvp.RecipesMvp;
import com.ladsoft.bakingapp.mvp.model.RecipeModel;
import com.ladsoft.bakingapp.mvp.model.RecipesModel;
import com.ladsoft.bakingapp.mvp.presenter.state.RecipeState;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public abstract class ModelModule {

    @Provides @Singleton
    static RecipesMvp.Model providesRecipesModel(RecipeRepository recipesRepository,
                                                 DatabaseRecipeRepository cacheRecipesRepository,
                                                 DatabaseIngredientRepository cacheIngredientRepository,
                                                 DatabaseStepRepository cacheStepRepository) {
        return new RecipesModel(recipesRepository,
                cacheRecipesRepository,
                cacheIngredientRepository,
                cacheStepRepository);
    }

    @Provides @Singleton
    static RecipeMvp.Model providesRecipeModel(DatabaseRecipeRepository cacheRecipesRepository,
                                               DatabaseIngredientRepository cacheIngredientRepository,
                                               DatabaseStepRepository cacheStepRepository) {
        return new RecipeModel(cacheRecipesRepository,
                cacheIngredientRepository,
                cacheStepRepository);
    }

    @Provides
    static RecipeState providesRecipeState(SessionManager sessionManager) {
        return new RecipeState(sessionManager);
    }
}

File: app/src/main/java/com/ladsoft/bakingapp/di/module/ModelModule.kt
package com.ladsoft.bakingapp.di.module

import com.ladsoft.bakingapp.data.database.repository.DatabaseIngredientRepository
import com.ladsoft.bakingapp.data.database.repository.DatabaseRecipeRepository
import com.ladsoft.bakingapp.data.database.repository.DatabaseStepRepository
import com.ladsoft.bakingapp.data.repository.RecipeRepository
import com.ladsoft.bakingapp.mvp.RecipeMvp
import com.ladsoft.bakingapp.mvp.RecipesMvp
import com.ladsoft.bakingapp.mvp.model.RecipeModel
import com.ladsoft.bakingapp.mvp.model.RecipesModel
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class ModelModule {

    @Provides @Singleton
    fun providesRecipesModel(recipesRepository: RecipeRepository,
                             cacheRecipesRepository: DatabaseRecipeRepository,
                             cacheIngredientRepository: DatabaseIngredientRepository,
                             cacheStepRepository: DatabaseStepRepository):
            RecipesMvp.Model = RecipesModel(recipesRepository,
                                    cacheRecipesRepository,
                                    cacheIngredientRepository,
                                    cacheStepRepository)

    @Provides @Singleton
    fun providesRecipeModel(cacheRecipesRepository: DatabaseRecipeRepository,
                             cacheIngredientRepository: DatabaseIngredientRepository,
                             cacheStepRepository: DatabaseStepRepository):
            RecipeMvp.Model = RecipeModel(cacheRecipesRepository,
                                        cacheIngredientRepository,
                                        cacheStepRepository)
}